feat: on demand config migration (#31129)
renovatebot/renovate: docs/usage/config-migration.md

While this feature is disabled by default, we strongly recommend enabling it to prevent unexpected behavior due to outdated configurations. Keeping your config up-to-date can significantly reduce debugging time for both users and maintainers when issues arise.

Here are the possible scenarios related to config migration:

1. **No config migration needed**

   Renovate takes no action.

2. **Config migration needed and enabled**

   Renovate creates a Config Migration PR and adds a link to it at the top of the Dependency Dashboard issue (if enabled).

3. **Config migration needed but disabled**

   If config migration is required but disabled:

   We add a checkbox to the Dependency Dashboard (if enabled). We call this "on-demand config migration".

   The checkbox will appear in your Dependency Dashboard issue as:

   ```
   [ ] Config migration needed. Please tick this checkbox to create an automated Config Migration PR.
   ```

   - When the user checks the box, Renovate creates a config migration PR and replaces the checkbox with a link:

   ```
   Config Migration necessary. You can view the Config Migration PR here #1

>    ```

Reviewer: Please update the section manually, so it looks like this:

### Config migration needed, but disabled

If config migration is needed, but disabled:

Renovate adds a checkbox to the Dependency Dashboard (if enabled).
This is "on-demand config migration".

The checkbox looks like this:

```
- [ ] Config migration needed. Select this checkbox to let Renovate create an automated Config Migration PR.
```

When you select the checkbox:

1. Renovate creates a config migration PR.
2. Renovate replaces the checkbox with text with a link to the Config Migration PR.

For example:

```
Config Migration needed. See Config Migration PR: #1.
```

> 4. **Config migration needed, but a closed migration PR exists**

Reviewer: ```suggestion
### Config migration needed, but there is a closed migration PR
```

>    In this case, regardless of whether config migration is enabled:
> 
>    - Renovate adds a checkbox to the Dependency Dashboard issue (if enabled).
>    - When checked, Renovate:
>      1. Deletes the old config migration branch.
>      2. Creates a new config migration PR.
>      3. Replaces the checkbox with a link to the new PR in the Dependency Dashboard issue.

Reviewer: ```suggestion
In this case, it does not matter if Config Migration is enabled, or not.
Renovate will:

- Add a checkbox to the Dependency Dashboard issue (if enabled)
- When you select the checkbox on the dashboard, Renovate will:
    1. Delete the _old_ config migration branch
    1. Create a _new_ config migration PR
    1. Replace the checkbox with a link to the _new_ PR in the Dependency Dashboard issue
```

<!-- prettier-ignore -->
!!! note

>     For config migration, Renovate's behavior differs from its usual approach to closed PRs. Instead of commenting on the closed PR that no new PR will be created. Instead Renovate adds a checkbox to the Dependency Dashboard issue, allowing users to request a new config migration PR when needed.

Reviewer: I'm not sure this is helpful. The text above already explains the behavior for Config Migration PRs:

1. Add checkbox to dashboard
2. Do thing when checkbox is selected

Also, I think Renovate should put a comment on closed/merged Config Migration PRs, that users should check:

- The new config migration PR?
- Or request a new config migration from the Dashboard?

Author: Yeah, this note maybe not be needed here.

We discussed in issue page to use a checkbox in the dashboard over the comment on PR.

# Config Migration

> Renovate maintainers often need to rename, remove or combine configuration options to improve the user experience and mature the product.

Reviewer: ```suggestion
To improve the Renovate program, the Renovate maintainers often rename, remove or combine configuration options.
```

> When they do so, "config migration" code is added at the same time so that any "legacy" config from users continues to work.

Reviewer: ```suggestion
When the Renovate maintainers change configuration options, they add "config migration" code.
The migration code allows "legacy" config from users to keep working.
```

> Config migration works by migrating legacy config internally before the config is used.

Reviewer: ```suggestion
Config migration works by migrating legacy config internally, before the config is used.
```

> If done right, it "just works" silently and legacy configs continue working indefinitely.

Reviewer: ```suggestion
If done right, config migration "just works" silently and legacy configs continue working indefinitely.
```

> The only sign when this is necessary is a debug message in logs noting the old and newly migrated configs.

Reviewer: ```suggestion
The only sign that "config migration" is needed is the debug message in the Renovate logs, noting the old and newly migrated configs.
```

> By default, none of these changes are applied to Renovate config files (e.g. `renovate.json`)

Reviewer: ```suggestion
By default, the "config migration needed" changes are applied at runtime, and then "forgotten".
By default, Renovate will _not_ update the Renovate config file itself (e.g. `renovate.json`).
```

How about something like this? The key point is that by _default_:

- Renovate silently applies the config migrations, and then "forgets about them"
- Renovate does not change/update the config file

Author: The "forgotten" word might confuse users. A simple `migrated at runtime` sounds better.

## Enabling config migration pull requests

> Although legacy config should continue to "just work", it's better for users if their config file uses the latest/correct configuration names.

Reviewer: ```suggestion
Even though Renovate allows you to keep using "legacy config", we recommend you update the configuration names in your config regularly.
```

> Using the latest names makes it easier to understand the config and look up documentation for it.

Reviewer: ```suggestion
Using the latest names:

- makes it easier for you to understand the config
- helps you find the documentation for the config
```

> Renovate can create a config migration pull request, that migrates legacy config in your configuration file.

Reviewer: ```suggestion
Renovate can create a config migration Pull Request, to migrate legacy config in your configuration file.
```

To get config migration pull requests from Renovate: set the [`configMigration`](./configuration-options.md#configmigration) config option to `true`.

Config migration PRs are disabled by default.
But we _strongly_ recommend you enable config migration PRs, because:

- the config migration PR "tells" you something changed
- up-to-date terms help you search the Renovate documentation
- up-to-date terms help you, and us, debug problems quicker

## Config migration scenarios

The scenarios for config migration are:

- No config migration needed
- Config migration needed, and enabled
- Config migration needed, but disabled
- Config migration needed, but there is a previously closed migration PR

### No config migration needed

Renovate takes no action.

### Config migration needed, and enabled

Renovate will:

1. Create a Config Migration PR
1. If the Dependency Dashboard issue is enabled then Renovate puts a link to the Config Migration PR on the dashboard

### Config migration needed, but disabled

> If config migration is needed, but disabled then Renovate adds a checkbox to the Dependency Dashboard if one exists.

Reviewer: It's not clear to me what `if one exists` refers to:

- Do you mean the Dashboard itself exists?
- Or do you mean that a "Config Migration PR" exists?

Author: I meant the Dashboard.

Maybe this works better: `if it exists`

This is known as "on-demand" config migration because migration PRs are only created at the request of the user by ticking the checkbox.

The checkbox looks like this:

```
- [ ] Select this checkbox to let Renovate create an automated Config Migration PR.
```

When you select the checkbox:

> Renovate creates a config migration PR.
> Renovate replaces the checkbox with a link to the Config Migration PR.

Reviewer: ```suggestion
1. Renovate creates a config migration PR
2. Renovate replaces the checkbox with a link to the Config Migration PR
```

# Config Migration

The developers may need to rename user-facing configuration options.
For example to:

- Use a better name for a config option, preset or concept
- Make a experimental configuration option officially supported

When this happens, the developers write "config migration" code to tell Renovate the old name and the new name.

When Renovate runs, it silently swaps old terms for new terms.
By default, Renovate does _not_ update the terms in your Renovate configuration file!

## Enabling config migration pull requests

Renovate can create a config migration pull request, that updates old terms in your configuration file.

> To enable config migration pull request from Renovate, set the [`configMigration`](./configuration-options.md#configmigration) config option to `true`.

Reviewer: ```suggestion
To get config migration pull requests from Renovate: set the [`configMigration`](./configuration-options.md#configmigration) config option to `true`.
```

Config migration PRs are disabled by default.
But we _strongly_ recommend you enable config migration PRs, because:

- the config migration PR "tells" you something changed
- up-to-date terms help you search the Renovate documentation
- up-to-date terms help you, and us, debug problems quicker

## Config migration scenarios

The scenarios for config migration are:

- No config migration needed
- Config migration needed, and enabled
- Config migration needed, but disabled
- Config migration needed, but there is a closed migration PR

### No config migration needed

Renovate takes no action.

### Config migration needed, and enabled

Renovate will:

1. Create a Config Migration PR
1. If the Dependency Dashboard issue is enabled: Renovate puts a link to the Config Migration PR on the dashboard

> ### Config migration needed but disabled

Reviewer: ```suggestion
### Config migration needed, but disabled
```

This is to match the list on line 32.

If config migration is needed, but disabled:

Renovate adds a checkbox to the Dependency Dashboard (if enabled).
This is "on-demand config migration".

The checkbox looks like this:

```
- [ ] Config migration needed. Select this checkbox to let Renovate create an automated Config Migration PR.
```

When you select the checkbox:

Renovate creates a config migration PR.

> Renovate replaces the checkbox with text with a link to the Config Migration PR.

Reviewer: ```suggestion
Renovate replaces the checkbox with a link to the Config Migration PR.
```

I think we can simplify, at the cost of a little bit of technical accuracy.

For example:

```
Config Migration needed. See Config Migration PR: #1.
```

### Config migration needed, but there is a closed migration PR

In this case, it does not matter if Config Migration is enabled, or not.
Renovate will:

- Add a checkbox to the Dependency Dashboard issue (if enabled)
- When you select the checkbox on the dashboard, Renovate will:
  1. Delete the _old_ config migration branch

>   1. Create a _new_ config migration PR

Reviewer 2: Won't this only happen if config migration is _enabled_?

Author: No, even if migration is disabled & a closed migration pr exists.

We want to create new migration PR if checkbox is ticked. [^`]

[^1]: https://github.com/renovatebot/renovate/issues/19783#issuecomment-2208125031

Author: No, even if migration is disabled & a closed migration pr exists.

We want to create new migration PR if checkbox is ticked. [^1]

[^1]: https://github.com/renovatebot/renovate/issues/19783#issuecomment-2208125031

Reviewer 2: Sorry,  I misread

# Config Migration

> To improve the Renovate program, the Renovate maintainers often rename, remove or combine configuration options.

Reviewer 2: ```suggestion
As part of continuous improvement and refinement, the Renovate maintainers often rename, remove or combine configuration options.
```

When the Renovate maintainers change configuration options, they add "config migration" code.
The migration code allows "legacy" config from users to keep working.
Config migration works by migrating legacy config internally, before the config is used.
If done right, config migration "just works" silently and legacy configs continue working indefinitely.
The only sign that "config migration" is needed is the debug message in the Renovate logs, noting the old and newly migrated configs.

> 
> By default, none of these changes are applied to Renovate config files (e.g. `renovate.json`)

Reviewer 2: ```suggestion
```

## Enabling config migration pull requests

Even though Renovate allows you to keep using "legacy config", we recommend you update the configuration names in your config regularly.
Using the latest names:

- makes it easier for you to understand the config
- helps you find the documentation for the config

Renovate can create a config migration Pull Request, to migrate legacy config in your configuration file.

> To get config migration pull requests from Renovate: set the [`configMigration`](./configuration-options.md#configmigration) config option to `true`.

Reviewer 2: ```suggestion
To get automated config migration pull requests from Renovate: set the [`configMigration`](./configuration-options.md#configmigration) config option to `true`.
```

Config migration PRs are disabled by default.

> But we _strongly_ recommend you enable config migration PRs, because:

Reviewer 2: ```suggestion
But we recommend you enable config migration PRs, because:
```
